services/kafka/config.go

```diff
@@ -82,58 +84,59 @@ type WriteTarget struct {
 	PartitionAlgorithm string
 }
 
-func (c Config) WriterConfig(diagnostic Diagnostic, target WriteTarget) (kafka.WriterConfig, error) {
+func (c Config) writerConfig(diagnostic Diagnostic, target WriteTarget) (*kafka.Config, error) {
+	cfg := kafka.NewConfig()
+
 	if target.Topic == "" {
-		return kafka.WriterConfig{}, errors.New("topic must not be empty")
+		return cfg, errors.New("topic must not be empty")
 	}
-	var balancer kafka.Balancer
+	var partitioner kafka.PartitionerConstructor
 	if target.PartitionById {
 		switch target.PartitionAlgorithm {
-		case "crc32c", "":
-			balancer = &kafka.CRC32Balancer{}
+		case "crc32", "":
+			// can't use kafka.NewCustomHashPartitioner, because we need to maintain compatibility
+			partitioner = newCRCPartitioner
 		case "murmur2":
-			balancer = &kafka.Murmur2Balancer{}
+			// can't use kafka.NewCustomHashPartitioner here either, because we need to maintain compatibility
+			partitioner = newMurmur2
+		case "murmur3":
+			partitioner = kafka.NewCustomHashPartitioner(murmur3.New32)
 		case "fnv-1a":
-			balancer = &kafka.Hash{}
+			partitioner = kafka.NewHashPartitioner
```

Comment:
Is this hasher also the same as before? I didn't see tests for that.

Reply:
Yes, the segmentio library claims it is the same.

Reply:
This test doesn't pass for me:

```
func partitions(n int) []int {
	p := make([]int, n)
	for i := range p {
		p[i] = i
	}
	return p
}

func TestHash(t *testing.T) {
	old := kafka.Hash{}
	newP := sarama.NewHashPartitioner

	for i := 0; i < 1000; i++ {
		partitionN := 10
		ps := partitions(partitionN)
		topic := fmt.Sprintf("foo-topic-%d", i)
		key := fmt.Sprintf("foo-key-%d", i)

		oldPartition := old.Balance(kafka.Message{
			Topic:     topic,
			Partition: 0,
			Offset:    0,
			Key:       []byte(key),
			Value:     nil,
			Time:      time.Time{},
		}, ps...)

		newPartitioner := newP(topic)
		newPartition, err := newPartitioner.Partition(&sarama.ProducerMessage{
			Topic: topic,
			Key:   sarama.StringEncoder(key),
		}, int32(len(ps)))
		require.NoError(t, err)

		assert.Equal(t, int(newPartition), oldPartition)
	}
}
```

Reply:
Then it was a bug in the old hasher, because we guarantee sarama compatibility.
https://github.com/influxdata/kapacitor/blob/4c2b9656fd7435c58ce1570c77058995b97b64f9/pipeline/alert.go#L2232

And in the docs.  
https://docs.influxdata.com/kapacitor/v1.6/event_handlers/kafka/#kafka-message-partitioning